flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java

```diff
@@ -247,11 +259,12 @@ public KubernetesWatch watchPodsAndDoCallback(
                                                                         new KubernetesPodsWatcher(
                                                                                 podCallbackHandler))),
                                         kubeClientExecutorService),
-                        maxRetryAttempts,
+                        new ExponentialBackoffRetryStrategy(
+                                maxRetryAttempts, initialRetryInterval, maxRetryInterval),
                         t ->
                                 ExceptionUtils.findThrowable(t, KubernetesClientException.class)
                                         .isPresent(),
-                        kubeClientExecutorService)
+                        new ScheduledExecutorServiceAdapter(kubeClientExecutorService))
                 .get();
```

Comment:
This means `watchPodsAndDoCallback` is blocking and can wait for minutes to return. It should not longer be called on the RPC main thread.